### DISCUSSION TITLE
Wrong value set for button in DataGrid binding when scrolling

### DISCUSSION THREAD

**Discussion by maguro1kan** [AvaloniaUI/Avalonia#12380] [General]
I have a issue of 'Wrong value set for button in DataGrid binding when scrolling'.

The correct value is retrieved from the first row up to the row to be scrolled, but after scrolling, a different value is retrieved.

Tell me how to fix it. Thanks.

Here is some codes
MainWindow.xaml
```
<StackPanel>
    <DataGrid Margin="20" 
              ItemsSource="{Binding Users}"
              GridLinesVisibility="All"
              IsReadOnly="True"
              Height="195"
              BorderThickness="1" 
              VerticalScrollBarVisibility="Visible"
              BorderBrush="Gray">
        <DataGrid.Columns>
            <DataGridTextColumn Header="UserID" Width="1.5*"
                                Binding="{Binding UserId}" />
<DataGridTemplateColumn Header="Action">
<DataGridTemplateColumn.CellTemplate>
<DataTemplate>
                        <Button x:Name="Check"
                                Name="{Binding UserId}"
                                Click="Check_OnClick"
                                Command="{Binding RelativeSource={RelativeSource AncestorType={x:Type UserControl}, Mode=FindAncestor}, Path=DataContext.CheckCommand}"
                                CommandParameter="{Binding ElementName=Check}">Check</Button>
</DataTemplate>
</DataGridTemplateColumn.CellTemplate>
</DataGridTemplateColumn>
        </DataGrid.Columns>
    </DataGrid>
    <TextBlock Padding="20 0 0 0">Log</TextBlock>
    <TextBox Margin="20 0" 
             x:Name="logTextBox"
             Height="180"
             Text="{Binding LogText}"
             AcceptsReturn="True"
             TextWrapping="Wrap"/>
</StackPanel>
```

MainWindow.xaml.cs
```
private async void Check_OnClick(object? sender, RoutedEventArgs e)
{
    string id = string.Empty;
    await Dispatcher.UIThread.InvokeAsync(() => { id = ((Button)sender).Name; }, DispatcherPriority.Background);
    Debug.Print(id);
    AddLog($"Clicked: {id}");
}

public void AddLog(string log)
{
     logTextBox.Text += log + Environment.NewLine;
}
```
![Desktop-2023 07 30-16 31 38 02](https://github.com/AvaloniaUI/Avalonia/assets/30754124/f01e9de9-b38b-413e-bb9b-ba81f08954c4)

**Comment by timunie** [comment]
Two things I see directly. 

1. Don't place ItemControls such as DataGrid inside a StackPanel. This will mess with virtualization. Better to limit it's height or place into a Grid or DockPanel. DataGrid has a ScrollViewer on its own.

2. You show the MainWindow code. However in one Binding you are looking for a UserControl parent which is not there. So please check that Binding. `$parent[Window].DataContext` should work here.

**Reply by timunie** [reply]
Okay forgot 1 as I have height overseen.

**Reply by maguro1kan** [reply]
Is this right?
```
  <Button x:Name="Check"
          Name="{Binding $parent[Window].DataContext.UserId}"
          Click="Check_OnClick">Check</Button>
```

I got a error.
`[Binding]Error in binding to 'Avalonia.Controls.Button'.'Name': 'Could not find a matching property accessor for 'UserId' on 'BugSample1.ViewModels.MainWindowViewModel''(Button #3865173)`

Part of MainWindowViewModel.cs 
```
  public class MainWindowViewModel : BindableBase
  {
      public MainWindowViewModel()
      {
          Users = new ObservableCollection<User>(Enumerable.Range(0, 14).Select(_ => new User()));
      }

      private ObservableCollection<User> _users;
      public ObservableCollection<User> Users
      {
          get => _users;
          set
          {
              _users = value;
              RaisePropertyChanged(nameof(Users));
          }
      }
  }

  public class User
  {
      public string UserId { get; set; }
      public string ClickedId { get; set; }

      public User()
      {
          UserId = Guid.NewGuid().ToString().ToUpper();
          ClickedId = "";
      }
  }
```

**Reply by timunie** [reply]
Woah I think there's something strange going on. Can you upload a simple sample to GitHub and file the link here? Will have  a look and send you a fix there hopefully im the next few days.

**Reply by maguro1kan** [reply]
Thanks!

https://github.com/maguro1kan/BugSample1

**Reply by timunie** [reply]
A couple of things: 
- `UserId` is a property of the item, so don't bind to winodows DataContext
- Name should not be bound to imo. Use `Tag` which works more reliable. Name is more to find the control in code. If name changes over and over again, this messes up as you see
- If you use FluentTheme, DataGrid also should be Fluent-Theme. 

Will send you a Fix to your sample.

**Reply by maguro1kan** [reply]
It worked!! Many thanks!
